api/envoy/extensions/filters/http/ratelimit/v3/rate_limit.proto

```diff
@@ -134,6 +134,26 @@ message RateLimit {
   // Optional additional prefix to use when emitting statistics. This allows to distinguish
   // emitted statistics between configured ``ratelimit`` filters in an HTTP filter chain.
   string stat_prefix = 13;
+
+  // If true, rate limit requests will also be sent to the rate limit service when the stream completes.
+  // This is useful when the rate limit budget needs to reflect the response context that is not available
+  // on the request path.
+  //
+  // On the stream completion, the filter will reuse the exact same descriptors matched during the request path.
+  // In other words, the descriptors are not recalculated on the stream completion, but the rate limit requests
+  // are sent with the same descriptors as the original request sent during the request path.
+  // For example, request header matching descriptors are available on the stream completion.
+  //
+  // For example, let's say the upstream service calculates the usage statistics, returns them in the response body
+  // and we want to utilize these numbers to apply the rate limit action for the subsequent requests.
+  // Combined with another filter that can set ``envoy.ratelimit.hits_addend`` based on the response (e.g. Lua filter),
+  // this can be used to subtract the usage statistics from the rate limit budget.
+  //
+  // The rate limit requests sent on the stream completion are "fire-and-forget" by nature, and rate limit is not enforced
+  // on the current HTTP stream being completed. The filter will only update the budget for the subsequent requests at
+  // that point. Hence the effect of the rate limit requests made during the stream completion is not visible in the current
+  // but only in the subsequent requests.
+  bool apply_on_stream_done = 14;
```

Comment:
1. Rather than a filter level flag, I think it should be per-descriptor level (in the route.v3.RateLimit). Then different routes could have different choice. And more important, different resource could also have different choice. Like the qps limit need only to works at request path but only the ai tokens limit need to works at response.

3. This semantics should be clear: request or response, but not for both. When we populate the descritpors, we can tell the the populateDescritors() what phase is and lets the populateDescritpors() only populate related descritpros. This bring a complexity that the users need to configure two descritpors: one for request with 0 or 1 hits addend for check only, one for response with CUSTOM_HITS_ADDEND or CUSTOM_HITS_ADDEND - 1 for report. But the semantics is more clear and behavior is more predictable. If we let same descritpor works on both request and response, we must ensure we can get correct different values for `hits_addend` from same source, that's weird.

Reply:
per descriptor flag sounds good. Any concerns? @arkodg

Reply:
reg 1. I dont understand per descriptor, per action makes sense (which generates a list of descriptors)
reg 2. and spitting up semantics into request or response makes it harder to build ratelimit actions e.g. the goal is user X (`x-user-id: X`) cannot exceed their usage by Y units/second. If two separate actions are used, then they'll need to generate the same descriptor list, imo its easier to generate the descriptor list using one action

Reply:
> reg 1. I dont understand per descriptor, per action makes sense (which generates a list of descriptors)

descriptor (conposite of multiple descriptor entries) is the minimal unit to match a rate limit rule. Eveny descriptor will be evaluated independently. Actually, I didn't get what you mean by per action? Single action should only generate a single descriptor entry, an `actions` (a list of single action) will generate a descriptor? Is there a info gap?

> reg 2. and spitting up semantics into request or response makes it harder to build ratelimit actions e.g. the goal is user X (x-user-id: X) cannot exceed their usage by Y units/second. If two separate actions are used, then they'll need to generate the same descriptor list, imo its easier to generate the descriptor list using one action which send the same descriptors in request and response path

This could be achieved by configure two `actions` that will generate same descriptor but has different hits_addend configuration. The complexity of configuration could be hide by the production encapsulation.

Independent `actions` means more flexible feature that we can use the filter to do only report without check or report usage for all accounts but only limit part of them, etc. And the more important is it make the API more clear and predictable. I can clear know that there are two descritors with different hits_addend, and will be applied at different phases.

Reply:
yeah descriptor is the unit of rate limiting which consists of multiple "descriptor entries" produced by actions. I think that's what arko and I got confused about it before. Now that we are trying to have a new descriptor level dynamic addend in https://github.com/envoyproxy/envoy/pull/37567, having descriptor level apply_on_stream_done would make sense.

Having two same action lists makes it feel a bit redundant, but having an explicit configuration for request and response separately I think is clearer.

Will make a change accordingly

Reply:
done https://github.com/envoyproxy/envoy/pull/37548/commits/57564f4525416df1ac7527ebc1585c5ce90d9c2a thank you for the comment @wbpcode